
# Ejercicio 2: Muertes por armas

El documento que se llama "guns.csv" contiene información sobre muertes por armas en EEUU. Vamos a crear funciones que nos ayuden a analizar los datos, a utilizar más de una base de datos y a practicar trabajar con fechas. Los datos son los siguientes: 

- ```year:``` Año en que ocurrió

- ```month:``` mes en que ocurrió

- ```intent:``` tipo de crimen

- ```police:``` si involucró un policía

- ```sex:``` genero de la víctima

- ```age:``` edad de la víctima

- ```race:``` raza de la víctima

- ```hispanic:``` indica el origen hispánico de la victima

- ```place:``` lugar donde ocurrió

- ```educación:``` 1: menos que secundaria, 2: graduado de secundaria, 3: algo de universidad, 4: graduado de universidad, 5: no se cuenta con la información



#### 1. Importar los datos
Crear una función que importe un csv, para este ejercicio puede utilizar la librería csv de Python. Para cargar  el directorio actual en el que nos encontramos podemos utilizar los "magic functions" de jupyter. En este caso si utilizamos %pwd podemos guardar el directorio en un avariable para utilizarlo después. 

In [1]:
import csv

In [3]:
path = r"./datasets/guns.csv"
def import_csv(path):
    with open(path,"r") as file:
        reader = csv.reader(file)
        raw_data = list(reader)
    return raw_data

#### 2. Separar los encabezados
Separar los encabezados del resto de los datos para usarlos para el análisis

In [6]:
def get_headers(raw_data):
    headers = raw_data[0]
    clean_data = raw_data[1:]
    return headers,clean_data

#### 3. Contar las muertes por año
Escriba una función que reciba una lista de datos y devuelva los conteos de cada ítem en la lista, calcule los conteos por año y los conteos por género y raza. 

In [11]:
def deaths_by_year(data):
    years = [row[1] for row in data]
    year_count = {}
    for yr in years:
        if yr not in year_count:
            year_count[yr] = 1
        else:
            year_count[yr] += 1
    return year_count

{'2012': 33563, '2013': 33636, '2014': 33599}

#### 4. Verificar si existen tendencias por género y raza
Describa en una celda de markup lo nota de los resultados obtenidos hasta el momento. 

In [18]:
def gender_tendency(data):
    genders = [row[5] for row in data]
    gender_counts = {}
    for gender in genders:
        if gender not in gender_counts:
            gender_counts[gender] = 0
        gender_counts[gender] += 1
    return gender_counts

In [21]:
def race_tendency(data):
    races = [row[7] for row in data]
    race_counts = {}
    for race in races:
        if race not in race_counts:
            race_counts[race] = 0
        race_counts[race] += 1
    return race_counts

##### Deaths by Year:
{'2012': 33563, '2013': 33636, '2014': 33599}
##### Gender Tendency:
{'M': 86349, 'F': 14449}
##### Race Tendency:
{'Asian/Pacific Islander': 1326, 'White': 66237, 'Native American/Native Alaskan': 917, 'Black': 23296, 'Hispanic': 9022}

#### 5. Formula para separar columnas

Crear una función que reciba los datos y devuelva una lista de columnas, esto corresponde a transponer los datos. 

In [34]:
def data_traspose(data):
    t_data = zip(data)
    f_data = []
    for x in t_data:
        f_data.append(x)
    return f_data

#### 6. Cargar datos de Censo

Crear otra variable que cargue los datos del documento "census.csv"

In [36]:
path = r"./datasets/census.csv"
census = import_csv(path)

#### 7. Eventos por cada 100 000 habitantes

Cree una función que calcule la proporción de eventos por raza por cada 100 000 habitantes. 

Primero creamos un diccionario de equivalencias apra que tenga sentido el cálculo:

````mapping = {
    'Asian/Pacific Islander': 674625 + 15159516 ,
     'Black': 40250635,
     'Hispanic': 44618105,
     'Native American/Native Alaskan': 3739506,
     'White': 197318956
}````


In [37]:
mapping = {
    'Asian/Pacific Islander': 674625 + 15159516 ,
     'Black': 40250635,
     'Hispanic': 44618105,
     'Native American/Native Alaskan': 3739506,
     'White': 197318956
}

#### 8. Filtrar por motivo
Crear una función que calcule las proporciones de crímines por cada 100 000 habitantes y que filtre el resultado por motivo. En otras palabras, que se presenten los datos por cada raza de el motivo del incidente (cantidad de personas blancas que cometieron un crimen de con el motivo de homicidio). 

In [ ]:
intent = set(intent)
print(intent)

In [ ]:
def crimes_per_capita(race):
    count_sui = 0
    count_hom = 0
    count_acc = 0
    count_und = 0
    count_na  = 0
    for obs in raw_data:
        if obs[7] == race:
            if obs[3] == 'Suicide':
                count_sui += 1
            elif obs[3] == 'Homicide':
                count_hom += 1
            elif obs[3] == 'Accidental':
                count_acc += 1
            elif obs[3] == 'Undetermined':
                count_und += 1
            elif obs[3] == 'NA':
                count_na += 1
    
    per_cap_sui = (count_sui / mapping[race]) * 100000
    per_cap_hom = (count_hom / mapping[race]) * 100000
    per_cap_acc = (count_acc / mapping[race]) * 100000
    per_cap_und = (count_und / mapping[race]) * 100000
    per_cap_na  = (count_na  / mapping[race]) * 100000
    
    print('Entre 2012 y 2014, por motivos de suicidio, en la raza {} murieron {} por cada 100.000 habitantes'.format(race,round(per_cap_sui,2)))
    print('Entre 2012 y 2014, por motivos de homicidio, en la raza {} murieron {} por cada 100.000 habitantes'.format(race,round(per_cap_hom,2)))
    print('Entre 2012 y 2014, por motivos accidentales, en la raza {} murieron {} por cada 100.000 habitantes'.format(race,round(per_cap_acc,2)))
    print('Entre 2012 y 2014, por motivos desconocidas, en la raza {} murieron {} por cada 100.000 habitantes'.format(race,round(per_cap_und,2)))
    print('Entre 2012 y 2014, no se tiene informacion, en la raza {} murieron {} por cada 100.000 habitantes'.format(race,round(per_cap_na,2)))
    
crimes_per_capita("Black")

####  9. Analizando más allá

Responda las siguientes preguntas con los datos, tods las respuestas deben venir acompañadas por una celda Markup en la que se explique su análisis.

##### a. ¿Existe una relación entre el mes en que se comete el crimen y la proporción de homicidios?

In [ ]:
months = set(month)
count_01 = 0
count_02 = 0
count_03 = 0
count_04 = 0
count_05 = 0
count_06 = 0
count_07 = 0
count_08 = 0
count_09 = 0
count_10 = 0
count_11 = 0
count_12 = 0

for mes in raw_data:
    count_01 = count_01 + 1 if mes[2] == '01' else count_01
    count_02 = count_02 + 1 if mes[2] == '02' else count_02
    count_03 = count_03 + 1 if mes[2] == '03' else count_03
    count_04 = count_04 + 1 if mes[2] == '04' else count_04
    count_05 = count_05 + 1 if mes[2] == '05' else count_05
    count_06 = count_06 + 1 if mes[2] == '06' else count_06
    count_07 = count_07 + 1 if mes[2] == '07' else count_07
    count_08 = count_08 + 1 if mes[2] == '08' else count_08
    count_09 = count_09 + 1 if mes[2] == '09' else count_09
    count_10 = count_10 + 1 if mes[2] == '10' else count_10
    count_11 = count_11 + 1 if mes[2] == '11' else count_11
    count_12 = count_12 + 1 if mes[2] == '12' else count_12
        
print("Durante los meses de enero se dieron {} muertes por armas de fuego".format(str(count_01)))
print("Durante los meses de febrero se dieron {} muertes por armas de fuego".format(str(count_02)))
print("Durante los meses de marzo se dieron {} muertes por armas de fuego".format(str(count_03)))
print("Durante los meses de abril se dieron {} muertes por armas de fuego".format(str(count_04)))
print("Durante los meses de mayo se dieron {} muertes por armas de fuego".format(str(count_05)))
print("Durante los meses de junio se dieron {} muertes por armas de fuego".format(str(count_06)))
print("Durante los meses de julio se dieron {} muertes por armas de fuego".format(str(count_07)))
print("Durante los meses de agosto se dieron {} muertes por armas de fuego".format(str(count_08)))
print("Durante los meses de septiembre se dieron {} muertes por armas de fuego".format(str(count_09)))
print("Durante los meses de octubre se dieron {} muertes por armas de fuego".format(str(count_10)))
print("Durante los meses de noviembre se dieron {} muertes por armas de fuego".format(str(count_11)))
print("Durante los meses de diciembre se dieron {} muertes por armas de fuego".format(str(count_12)))

##### b. ¿Existe una correlación entre las razones de muertes por armas y la ubicación (location)? (Muestre su análisis con código y escriba en una celda sus observaciones)

In [ ]:
place = set(place)
print(place)
count_farm     = 0
count_oth_spec = 0
count_street   = 0
count_home     = 0
count_res      = 0
count_sports   = 0
count_na       = 0
count_oth_unsp = 0
count_school   = 0
count_trade    = 0
count_ind      = 0

for lugar in raw_data:
    count_farm     = count_farm     + 1 if lugar[9] == 'Farm' else count_farm
    count_oth_spec = count_oth_spec + 1 if lugar[9] == 'Other specified' else count_oth_spec
    count_street   = count_street   + 1 if lugar[9] == 'Street' else count_street
    count_home     = count_home     + 1 if lugar[9] == 'Home' else count_home
    count_res      = count_res      + 1 if lugar[9] == 'Residential institution' else count_res
    count_sports   = count_sports   + 1 if lugar[9] == 'Sports' else count_sports
    count_na       = count_na       + 1 if lugar[9] == 'NA' else count_na
    count_oth_unsp = count_oth_unsp + 1 if lugar[9] == 'Other unspecified' else count_oth_unsp
    count_school   = count_school   + 1 if lugar[9] == 'School/instiution' else count_school
    count_trade    = count_trade    + 1 if lugar[9] == 'Trade/service area' else count_trade
    count_ind      = count_ind      + 1 if lugar[9] == 'Industrial/construction' else count_ind
    
print("En total se dieron {} muertes por armas de fuego en granjas".format(str(count_farm)))
print("En total se dieron {} muertes por armas de fuego en otros lugares".format(str(count_oth_spec)))
print("En total se dieron {} muertes por armas de fuego en calles".format(str(count_street)))
print("En total se dieron {} muertes por armas de fuego en casas".format(str(count_home)))
print("En total se dieron {} muertes por armas de fuego en residencias".format(str(count_res)))
print("En total se dieron {} muertes por armas de fuego en lugares deportivos".format(str(count_sports)))
print("En total se dieron {} muertes por armas de fuego tuvieron un sitio no especificado".format(str(count_na)))
print("En total se dieron {} muertes por armas de fuego en lugares no especificados".format(str(count_oth_unsp)))
print("En total se dieron {} muertes por armas de fuego en escuelas".format(str(count_school)))
print("En total se dieron {} muertes por armas de fuego en lugares de comercio".format(str(count_trade)))
print("En total se dieron {} muertes por armas de fuego en parques industriales".format(str(count_ind)))

##### c. ¿Existe una correlación entre las razones de muertes por armas y el nivel educativo?  (Muestre su análisis con código y escriba en una celda sus observaciones)

In [ ]:
print(set(education))
count_ba_plus  = 0
count_hs_ged   = 0
count_some_c   = 0
count_less_hs  = 0
count_na       = 0

for educ in raw_data:
    count_ba_plus = count_ba_plus + 1 if educ[10] == 'BA+' else count_ba_plus
    count_hs_ged  = count_hs_ged + 1 if educ[10] == 'HS/GED' else count_hs_ged
    count_some_c  = count_some_c + 1 if educ[10] == 'Some college' else count_some_c
    count_less_hs = count_less_hs + 1 if educ[10] == 'Less than HS' else count_less_hs
    count_na      = count_na + 1 if educ[10] == 'NA' else count_na
    
print("BA+ : {}".format(str(count_ba_plus)))
print("HS/GED : {}".format(str(count_hs_ged)))
print("Some college : {}".format(str(count_some_c)))
print("Less than HS : {}".format(str(count_less_hs)))
print("NA : {}".format(str(count_na)))

##### d. ¿Cuál es la razón de homicidios por género?

In [ ]:
count_male = 0
count_female = 0

for sex in raw_data:
    count_male = count_male + 1 if sex[5] == "M" else count_male
    count_female = count_female + 1 if sex[5] == "F" else count_female

ratio = count_male / count_female

print("Por cada mujer que fallece por armas de fuego {0:.2f} hombres mueren".format(float(ratio)))

##### e. ¿Cuál es la relación, si es que hay, entre la categoría del incidente y raza? Por ejemplo los incidentes accidentales son más propensos en alguna raza?

In [ ]:
print(set(race))
print(set(intent))

count_w_s = 0
count_w_h = 0
count_w_a = 0
count_w_u = 0
count_w_n = 0

count_b_s = 0
count_b_h = 0
count_b_a = 0
count_b_u = 0
count_b_n = 0

count_n_s = 0
count_n_h = 0
count_n_a = 0
count_n_u = 0
count_n_n = 0

count_h_s = 0
count_h_h = 0
count_h_a = 0
count_h_u = 0
count_h_n = 0

count_a_s = 0
count_a_h = 0
count_a_a = 0
count_a_u = 0
count_a_n = 0

for i in raw_data:
    if i[7] == "White" and i[3] == "Suicide":
        count_w_s += 1
    elif i[7] == "White" and i[3] == "Homicide":
        count_w_h += 1
    elif i[7] == "White" and i[3] == "Accidental":
        count_w_a += 1
    elif i[7] == "White" and i[3] == "Undetermined":
        count_w_u += 1
    elif i[7] == "White" and i[3] == "NA":
        count_w_n += 1
    elif i[7] == "Black" and i[3] == "Suicide":
        count_b_s += 1
    elif i[7] == "Black" and i[3] == "Homicide":
        count_b_h += 1
    elif i[7] == "Black" and i[3] == "Accidental":
        count_b_a += 1
    elif i[7] == "Black" and i[3] == "Undetermined":
        count_b_u += 1
    elif i[7] == "Black" and i[3] == "NA":
        count_b_n += 1   
    elif i[7] == "Native American/Native Alaskan" and i[3] == "Suicide":
        count_n_s += 1
    elif i[7] == "Native American/Native Alaskan" and i[3] == "Homicide":
        count_n_h += 1
    elif i[7] == "Native American/Native Alaskan" and i[3] == "Accidental":
        count_n_a += 1
    elif i[7] == "Native American/Native Alaskan" and i[3] == "Undetermined":
        count_n_u += 1
    elif i[7] == "Native American/Native Alaskan" and i[3] == "NA":
        count_n_n += 1
    elif i[7] == "Hispanic" and i[3] == "Suicide":
        count_h_s += 1
    elif i[7] == "Hispanic" and i[3] == "Homicide":
        count_h_h += 1
    elif i[7] == "Hispanic" and i[3] == "Accidental":
        count_h_a += 1
    elif i[7] == "Hispanic" and i[3] == "Undetermined":
        count_h_u += 1
    elif i[7] == "Hispanic" and i[3] == "NA":
        count_h_n += 1
    elif i[7] == "Asian/Pacific Islander" and i[3] == "Suicide":
        count_a_s += 1
    elif i[7] == "Asian/Pacific Islander" and i[3] == "Homicide":
        count_a_h += 1
    elif i[7] == "Asian/Pacific Islander" and i[3] == "Accidental":
        count_a_a += 1
    elif i[7] == "Asian/Pacific Islander" and i[3] == "Undetermined":
        count_a_u += 1
    elif i[7] == "Asian/Pacific Islander" and i[3] == "NA":
        count_a_n += 1
        
        
print(count_n_s)
print(count_n_h)
print(count_n_a)
print(count_n_u)
print(count_n_n)

In [ ]:
print("                         |  Suicide | Homicide | Accidental | Undetermined |    NA   |")
print("-------------------------|----------|----------|------------|--------------|---------|")
print(" Native American/Alaskan | {}      | {}      | {}         | {}           | {}       |".format(int(count_n_s),int(count_n_h),int(count_n_a),int(count_n_u),int(count_n_n)))
print("-------------------------|----------|----------|------------|--------------|---------|")
print("        White            | {}    | {}     | {}       | {}          | {}       |".format(int(count_w_s),int(count_w_h),int(count_w_a),int(count_w_u),int(count_w_n)))
print("-------------------------|----------|----------|------------|--------------|---------|")
print("        Black            | {}     | {}    | {}        | {}          | {}       |".format(int(count_b_s),int(count_b_h),int(count_b_a),int(count_b_u),int(count_b_n)))
print("-------------------------|----------|----------|------------|--------------|---------|")
print("        Hispanic         | {}     | {}     | {}        | {}           | {}       |".format(int(count_h_s),int(count_h_h),int(count_h_a),int(count_h_u),int(count_h_n)))
print("-------------------------|----------|----------|------------|--------------|---------|")
print(" Asian/Pacific Islander  | {}      | {}      | {}         | {}           | {}       |".format(int(count_a_s),int(count_a_h),int(count_a_a),int(count_a_u),int(count_a_n)))